engine/runtime/src/main/java/org/enso/interpreter/runtime/builtin/Error.java

```diff
@@ -207,4 +219,135 @@ public Atom makeModuleDoesNotExistError(String name) {
   public Atom makeNotInvokableError(Object target) {
     return notInvokableError.newInstance(target);
   }
+
+  /** Represents plain Java exception as a {@link TruffleObject}.
+   */
+  @ExportLibrary(InteropLibrary.class)
+  static final class WrapPlainException extends AbstractTruffleException {
+    private final AbstractTruffleException prototype;
+    private final Throwable original;
+
+    private WrapPlainException(Throwable cause) {
+      super(cause.getMessage(), cause, AbstractTruffleException.UNLIMITED_STACK_TRACE, null);
+      this.prototype = null;
+      this.original = cause;
+    }
+
+    private WrapPlainException(AbstractTruffleException prototype, Throwable original) {
+      super(prototype);
+      this.prototype = prototype;
+      this.original = original;
+    }
+
+    static AbstractTruffleException wrap(Throwable cause, Context ctx) {
+      var env = ctx.getEnvironment();
+      if (env.isHostException(cause)) {
+        var orig = env.asHostException(cause);
+        return new WrapPlainException((AbstractTruffleException) cause, orig);
+      } else if (cause instanceof AbstractTruffleException truffleEx) {
+        return truffleEx;
+      } else {
+        return new WrapPlainException(cause);
+      }
+    }
+
+    @ExportMessage
+    boolean hasExceptionMessage() {
+      return getMessage() != null;
+    }
+
+    @ExportMessage
+    public Object getExceptionMessage() {
+      return Text.create(getMessage());
+    }
+
+    @ExportMessage
+    String toDisplayString(boolean sideEffects) {
+      return original.toString();
+    }
+
+    @ExportMessage
+    Object getMembers(boolean includeInternal) {
+      return Array.empty();
+    }
+
+    @ExportMessage
+    boolean hasMembers() {
+      return true;
+    }
+
+    @ExportMessage
+    boolean isMemberInvocable(String member, @CachedLibrary(limit="1") InteropLibrary delegate) {
+      if (
+        "has_type".equals(member) ||
```

Comment:
I'd expect a piece of documentation which describes how `has_type` works (unless I missed it) and how its semantics differ from `Java.is_instance`. As otherwise, I don't understand why the change is made and what it means.

Reply:
As we have discussed with @JaroslavTulach, there seems to be no good place to document this `has_type` method, because it is added essentially to any Java type extending `Throwable`.
But IIRC only if the Throwable is caught, ones created with `new` may not necessarily have it - @JaroslavTulach is that the case? That's how I currently understand the wrapping mechanism - that it happens only at 'catch-site' - but I may be wrong. Thus, that may too be worth clarifying.

Thus, there is no Enso supertype common for exceptions that we could attach the doc to.

Because of that, we have decided the best course of action may be to add a `has_type` (or a similar name) method to the `type Polyglot_Error cause` Atom which would then delegate to the underlying wrapped exception's `has_type` implementation. Then the base `has_type` extension will not be directly documented, but all of its behaviour can be explained in the `Polyglot_Error.has_type` "facade".

Does that sound good?

Reply:
> because it is added essentially to any Java type extending Throwable

No, not really. One has to construct `Polyglot_Error` atom via `Error.makePolyglotError` - then the first argument is `AbstractTruffleException` instance which responds to `has_type` `invokeMember` message.

> there is no Enso supertype common for exceptions that we could attach the doc to.

No, there is not any common super type. The first argument of `Polyglot_Error` is of type `Dynamic`.

`Polyglot_Error.has_type` sounds OKish.

Reply:
At the end I got a better idea: [ec0b51d](https://github.com/enso-org/enso/pull/3559/commits/ec0b51d7a0058cadabb7e3aa722042c104bd5e3e) - there is `Any.is_a`, so let's reuse that method!